Stats helper

sainthkh opened the pull request:
Closes #12031 

Added `{{@stats}}` helper for users to write more persuasive messages on their sites.

### Todo

- [x] Retrieve related data from the database.
- [x] Refactor stats service to use a single file. `cache` is a bit an overkill.
- [x] Send data to templates by `hbs.updateTemplateOptions`
- [x] Add an unit test to `middleware_spec.js`
- ~~Don't cache `site_age` and `site_age_years` -> They should always be calculated when the middleware is called.~~
- ~~Update stats(posts, authors, tags, members_posts, paid_members_posts) when a post is created/updated/deleted.~~
- ~~Update stats(members) when a member is created/updated/deleted.~~

In current code, updated stats are not reflected to TemplateOptions. They're only set up once. Check the code below.

https://github.com/TryGhost/Ghost/blob/2d2fa1a0bae4129c72755810fb96be2d3feba508/core/frontend/services/themes/middleware.js#L87-L90

That's why I stroke out 3 todos. I guess this update feature should be a part of the following PR.

Please include a description of your change & check your PR against this list, thanks!

- [x] There's a clear use-case for this code change. => They're useful for upsell messages.
- [x] Commit message has a short title & references relevant issues
- [x] The build will pass (run `yarn test` and `yarn lint`)

More info can be found by clicking the "guidelines for contributing" link above.

sainthkh committed:
Add new stats service.

issue #12031

Added new stats service that retrieves data from the database.

sainthkh committed:
Refactor stats.

issue #12031

We don't need 2 files(index.js, cache.js) for this code size.

sainthkh committed:
Set stats at templateOptions.

issue #12031

To make {{@stats}} accessible in themes.

sainthkh committed:
Add test.

issue #12031

sainthkh committed:
Extract function.

issue #12031

naz commented on core/server/services/stats/index.js:
```diff
@@ -0,0 +1,35 @@
+    const createdAt = result.models[0].get('created_at');
+
+    const stats = {
+        total_members: await models.Member.count(),
```
This operation might become time consuming, as well as the rest of `.count()` calls, for large datasets. We'd need performance numbers to evaluate if it's worth keeping Bookshelf model call (suspecting it's not) or maybe drop down a level and query directly through knex. 

To start with performance metrics, you would need to populate the db with dummy members (don't have a public script that I could share :grimacing:). Would need to generate perf numbers for sets of 10K, 50K, 100K members.  Then would measure how long it takes to execute `models.Member.count()` for each set.

For example through `consoleTime` (or [hook into](https://spin.atomicobject.com/2017/03/27/timing-queries-knexjs-nodejs/) bookshelf model directly):
```
console.time('memberCount');

await models.Member.count()

console.timeEnd('memberCount');
```

Similar metrics but with lower numbers (5K, 10K, 30K) would need to be done for posts.

These numbers would lead to further decision making :crossed_fingers:

sainthkh replied:
Here are the numbers for Members. I've run the query 50 times and calculated the numbers.

| | 10K model | 10K knex |
| - | ---- | ---- |
| Avg | 31.481ms | 26.795ms |
| Min | 20.762ms 2nd | 9.868ms 1st |
| Max | 43.019ms 50th | 40.010ms 50th |

| | 50K model | 50K knex |
| - | ---- | ---- |
| Avg | 39.171ms | 28.613ms |
| Min | 23.665ms 2nd | 11.960ms 1st |
| Max | 57.316ms 50th | 43.915ms 50th |

I haven't checked posts and tags. But I believe the result would be similar. 

According to the test result, I suggest using knex directly is better for performance than using bookshelf models. 

Particularly, the first run of knex is about 2x faster than model.

If you want more numbers, I'll test it and send them to you.

sainthkh replied:
I've shared the tool I used to seed 50K data at #12272.

naz commented on core/server/index.js:
```diff
@@ -48,6 +49,7 @@ function initialiseServices() {
         mega.listen(),
         webhooks.listen(),
         appService.init(),
+        stats.init(),
```
Apart from this place, we'd also need to hook into other model specific events (like adding/removing a member or post) to keep stat numbers fresh.

sainthkh replied:
Inside `updateGlobalTemplateOptions()` function, there is a todo like this:

https://github.com/TryGhost/Ghost/blob/b34f2e86fb9a785e7483f6ffc4a1b0269f9f6709/core/frontend/services/themes/middleware.js#L100

When I first wrote this code, I thought solving this problem would make this PR too big. But it seems that it wouldn't do so. I'll handle this.

naz commented on test/unit/services/themes/middleware_spec.js:
```diff
@@ -5,6 +5,7 @@ const themes = require('../../../../core/frontend/services/themes');
 // is only exposed via themes.getActive()
 const activeTheme = require('../../../../core/frontend/services/themes/active');
 const settingsCache = require('../../../../core/server/services/settings/cache');
+const stats = require('../../../../core/server/services/stats');
```
Would need some more unit tests covering new helper itself, similarly to [these tests](https://github.com/TryGhost/Ghost/blob/b34f2e86fb9a785e7483f6ffc4a1b0269f9f6709/test/unit/helpers/pagination_spec.js).

sainthkh replied:
It seems that the wrong example is given. `{{@stats}}` is an global object, not helper functions. The pagination example and other tests inside that folder are about testing helper functions in `core/frontend/helpers`, not the global objects passed in `frontend/services/themes/middleware.js`. 

It does have the cases that use global objects like `@site` in `test/unit/helpers/concat_spec.js`, but they're testing `has` and `@site` is set up as `defaultGlobals` in `before`. The values are loaded from the different route. 

I think the best alternative is to use `test-theme` in the fixture or create a new theme to use `{{@stats}}` helper. Then, write an acceptance test that activates that theme and load the page to check if `{{@stats}}` are loaded correctly. 

But it sounds a bit big. Do you want me to do that?

sainthkh replied:
I slept on it for days and decided that it's the only solution we have for this case and the case for updating stats when something changed. 

I'll go down this road tomorrow.

naz reviewed (changes requested):
This is good direction, thanks for taking up on the issue! Let me know if I could help moving this issue forward :wink:

sainthkh commented on core/server/services/stats/index.js:
```diff
@@ -0,0 +1,35 @@
+
+async function initStats() {
+    const result = await models.Post.query(qb => qb.orderBy('created_at', 'ASC').limit(1)).fetchAll({columns: ['created_at']});
+    const createdAt = result.models[0].get('created_at');
```
Right. But the problem is, when should we set the `createdAt`? If we set it to `null`, it can break the theme.

I think there are options:

* current time: it's simple and fast (no more query). But in this case, the name and the value doesn't go well together.
* the oldest page time: if the blog has no page again, we encounter the same problem. 
* the timestamp of `db_hash` value at `settings` table: I think it's the best alternative because it's created when we set up a Ghost blog and it is unlikely that it would be deleted by the system. (The only possibility is the malicious attacks or the mistake by the user.) 

I think `db_hash` is the best fallback for the created date of the first post. What do you think?

naz replied:
Have asked a question in the main issue - https://github.com/TryGhost/Ghost/issues/12031#issuecomment-703490561. Let's see what the outcome is.

naz replied:
@sainthkh fyi the question above has been answered [here](https://github.com/TryGhost/Ghost/issues/12031#issuecomment-704972385). Hopefully that unblocks next steps in the implementation ;)

sainthkh committed:
Update stats when things changed.

issue #12031

sainthkh committed:
Return 0 for site_age when there is no post.

issue #12031

sainthkh commented:
Todos

- [x] Prevent post failure.
- [x] Extract function
- [ ] Count performance issue => Waiting for your approvement.
- [x] Update handlebars global options when stats changed.
- [ ] Add tests => Waiting for your opinion.

sainthkh committed:
Fix test failure.

sainthkh committed:
Add tests.

sainthkh committed:
Remove post.deleted event from the list.

It's because post.unpublished is called when deleting post. We don't have to count posts twice.

sainthkh committed:
Support free_member, paid_member counts.

sainthkh committed:
Support updating members/paid post counts.

sainthkh committed:
Add tag events.

sainthkh committed:
Add author tests.

sainthkh committed:
Add age tests.

sainthkh committed:
Fix test errors.

sainthkh committed:
Refactor stats.js

sainthkh committed:
Fix test failures and warnings.

sainthkh committed:
Fix foreign key errors.

sainthkh committed:
Don't emit event to update global template options.

sainthkh committed:
Fix failed unit test.

sainthkh committed:
Fix flakiness.

sainthkh committed:
Remove unnecessary request.

sainthkh committed:
Fix failed tests.

sainthkh committed:
Fix flakiness.

sainthkh commented on core/server/services/stats/init.js:
```diff
@@ -0,0 +1,107 @@
+async function updatePosts() {
+    cache.set('total_tags', await countTags());
+    cache.set('total_authors', await countAuthors());
+    cache.set('total_posts', (await db.knex('posts').where({status: 'published'}).count())[0]['count(*)']);
+    cache.set('total_members_posts', (await db.knex('posts').where({visibility: 'members', status: 'published'}).count())[0]['count(*)']);
+    cache.set('total_paid_members_posts', (await db.knex('posts').where({visibility: 'paid', status: 'published'}).count())[0]['count(*)']);
```
It's moved down because there is a scheduling issue. If we move them to the top of the function, `total_members_posts` doesn't work properly in MySQL.

sainthkh commented on test/regression/services/stats_spec.js:
```diff
@@ -0,0 +1,603 @@
+    return _.clone(testUtils.DataGenerator.forModel.posts[0]);
+}
+
+const PREVENT_FLAKINESS_TIME = 400;
```
I tried to reduce the number. But it causes flakiness. 

Maybe we can add an event at the end of the `updateGlobalTemplateOptions` and use it in the test.

sainthkh reviewed (commented):
@naz All things done. But there are 2 known issues.

ErisDS commented:
@sainthkh Thank you for this PR! I apologise for the delay but this is on my list to review ASAP.
